pkg/deploy/lattice/service_network_manager.go

```diff
@@ -274,22 +287,84 @@ func (m *defaultServiceNetworkManager) isServiceNetworkAssociatedWithVPC(ctx con
 					glog.V(6).Infoln("Mesh and Vpc association failed to delete")
-					return true, r.Id, resp, nil
+					return true, r, resp, nil
 				case vpclattice.ServiceNetworkVpcAssociationStatusDeleteInProgress:
 					glog.V(6).Infoln("ServiceNetwork and Vpc association is being deleted, please retry later")
-					return true, r.Id, resp, errors.New(LATTICE_RETRY)
+					return true, r, resp, errors.New(LATTICE_RETRY)
 				case vpclattice.ServiceNetworkVpcAssociationStatusCreateInProgress:
 					glog.V(6).Infoln("ServiceNetwork and Vpc association is being created, please retry later")
-					return true, r.Id, resp, errors.New(LATTICE_RETRY)
+					return true, r, resp, errors.New(LATTICE_RETRY)
 				}
 			}
 		}
 	}
 	return false, nil, resp, err
 }
+
+func (m *defaultServiceNetworkManager) UpdateServiceNetworkVpcAssociationIfNeeded(ctx context.Context, existingSN *vpclattice.ServiceNetworkSummary, desiredSN *latticemodel.ServiceNetwork, existingSnvaId *string) (latticemodel.ServiceNetworkStatus, error) {
+	retrievedSnva, err := m.cloud.Lattice().GetServiceNetworkVpcAssociationWithContext(ctx, &vpclattice.GetServiceNetworkVpcAssociationInput{
+		ServiceNetworkVpcAssociationIdentifier: existingSnvaId,
+	})
+	if err != nil {
+		return latticemodel.ServiceNetworkStatus{}, err
+	}
+
+	if len(desiredSN.Spec.SecurityGroupIds) == 0 && retrievedSnva.SecurityGroupIds != nil && len(retrievedSnva.SecurityGroupIds) > 0 {
+		// VPC lattice API have a limitation that the user can't remove all security groups if current snva already has security groups.
+		// If that case happen, the controller will not delete snva. User should manually delete it.
+		//https://docs.aws.amazon.com/vpc-lattice/latest/ug/security-groups.html#security-groups-rules
+		gwlog.FallbackLogger.Errorf("Can't remove all security groups from service network vpc association, for service network %s, snvaId %s", desiredSN.Spec.Name, retrievedSnva.Id)
+
+		// TODO: Add validation webhook or update gateway (or VpcAssociationPolicy) status to indicate that error.
+		//  https://gateway-api.sigs.k8s.io/geps/gep-713/#standard-status-condition
+		return latticemodel.ServiceNetworkStatus{}, errors.New(LATTICE_RETRY)
+	}
+
+	sgIdsEqual := securityGroupIdsEqual(desiredSN.Spec.SecurityGroupIds, retrievedSnva.SecurityGroupIds)
+
+	if sgIdsEqual {
+		// desiredSN's security group ids are same with retrievedSnva's security group ids, don't need to update
+		return latticemodel.ServiceNetworkStatus{
+			ServiceNetworkID:     *existingSN.Id,
+			ServiceNetworkARN:    *existingSN.Arn,
+			SnvaSecurityGroupIds: retrievedSnva.SecurityGroupIds,
+		}, nil
+	}
+
+	updateSnvaResp, err := m.cloud.Lattice().UpdateServiceNetworkVpcAssociationWithContext(ctx, &vpclattice.UpdateServiceNetworkVpcAssociationInput{
+		ServiceNetworkVpcAssociationIdentifier: existingSnvaId,
+		SecurityGroupIds:                       desiredSN.Spec.SecurityGroupIds,
+	})
+	if err != nil {
+		return latticemodel.ServiceNetworkStatus{}, err
+	}
```

Comment:
few questions:
1. Why do we need to handle "remove all security groups". I believe lattice API will throw error if we try to do so. I would let it go through and fail. If one day we remove this limitation from Lattice this code will fail for no reason.
2. If there is no diff in security group ids, does lattice do anything if still make a update call? If not I will also remove diff check here and call lattice, and let lattice figure-out diff.

Reply:
Will do some manual test for 1 and 2. For 2, I think the short circuit logic may needed since UpdateSNVA api has a pretty low throttling limit

Reply:
Cant relate low throughput and duplicating validation logic. We should not duplicate logic/code that exists in Lattice.

Reply:
Will remove the `if len(desiredSN.Spec.SecurityGroupIds) == 0 && retrievedSnva.SecurityGroupIds != nil && len(retrievedSnva.SecurityGroupIds) > 0 {}`  block in the next revision, from my test, lattice api already have a validation to prevent to update to a empty sgIds